File: packages/sdk-python/src/agentshield/session.py
"""ShieldSession — async context manager for guarded agent sessions."""

from __future__ import annotations

from typing import Any, Callable, Coroutine

from agentshield.client import ServerClient
from agentshield.exceptions import ConfirmationRejected, ToolCallBlocked
from agentshield.models import CheckResult, Decision


class GuardedExecutor:
    """Wraps arbitrary async tool calls through the shield check pipeline."""

    def __init__(
        self,
        client: ServerClient,
        session_id: str,
        confirm_callback: Callable[[str, dict], Coroutine[Any, Any, bool]] | None = None,
    ) -> None:
        self._client = client
        self._session_id = session_id
        self._confirm_callback = confirm_callback

    async def execute(
        self,
        tool_name: str,
        params: dict,
        func: Callable[..., Coroutine[Any, Any, Any]],
        *,
        source_id: str = "",
    ) -> Any:
        """Check with the server, enforce the decision, then run the tool."""
        result: CheckResult = await self._client.check_tool_call(
            session_id=self._session_id,
            tool_name=tool_name,
            params=params,
            source_id=source_id,
        )

        if result.action is Decision.BLOCK:
            raise ToolCallBlocked(
                tool=tool_name,
                reason=result.reason,
                trace_id=result.trace_id,
            )

        if result.action is Decision.REQUIRE_CONFIRMATION:
            if self._confirm_callback is None:
                raise ConfirmationRejected(tool=tool_name)
            confirmed = await self._confirm_callback(tool_name, params)
            if not confirmed:
                raise ConfirmationRejected(tool=tool_name)

        return await func(**params)


class ShieldSession:
    """Async context manager representing a guarded agent session.

    Usage::

        async with shield.session("Summarize my emails") as s:
            result = await s.guarded_executor.execute(
                "read_inbox", {"limit": 10}, read_inbox_fn
            )
    """

    def __init__(
        self,
        client: ServerClient,
        user_message: str,
        agent_id: str = "",
        metadata: dict | None = None,
        confirm_callback: Callable[[str, dict], Coroutine[Any, Any, bool]] | None = None,
    ) -> None:
        self._client = client
        self._user_message = user_message
        self._agent_id = agent_id
        self._metadata = metadata or {}
        self._confirm_callback = confirm_callback
        self.session_id: str = ""
        self.trace_id: str = ""
        self._executor: GuardedExecutor | None = None

    async def __aenter__(self) -> ShieldSession:
        info = await self._client.create_session(
            user_message=self._user_message,
            agent_id=self._agent_id,
            metadata=self._metadata,
        )
        self.session_id = info.session_id
        self.trace_id = info.trace_id
        self._executor = GuardedExecutor(
            client=self._client,
            session_id=self.session_id,
            confirm_callback=self._confirm_callback,
        )
        return self

    async def __aexit__(self, exc_type: type | None, exc_val: BaseException | None, exc_tb: object) -> None:
        # Session lifecycle is tracked server-side; nothing to tear down locally.
        pass

    @property
    def guarded_executor(self) -> GuardedExecutor:
        """Return the executor that routes tool calls through the server check."""
        if self._executor is None:
            raise RuntimeError("ShieldSession must be used as an async context manager")
        return self._executor

File: packages/sdk-typescript/src/session.ts
/** ShieldSession — guarded agent session with tool call interception. */

import {
  ConfirmationRejected,
  ServerClient,
  ToolCallBlocked,
} from "./client.js";
import type { CheckResult, ShieldConfig } from "./models.js";
import { Decision } from "./models.js";

export type ConfirmCallback = (
  toolName: string,
  params: Record<string, unknown>,
) => Promise<boolean>;

/**
 * Wraps arbitrary tool calls through the shield check pipeline.
 */
export class GuardedExecutor {
  constructor(
    private readonly client: ServerClient,
    private readonly sessionId: string,
    private readonly confirmCallback?: ConfirmCallback,
  ) {}

  /**
   * Check with the server, enforce the decision, then run the tool function.
   */
  async execute<T>(
    toolName: string,
    params: Record<string, unknown>,
    fn: (params: Record<string, unknown>) => Promise<T>,
    opts?: { sourceId?: string },
  ): Promise<T> {
    const result: CheckResult = await this.client.checkToolCall({
      sessionId: this.sessionId,
      toolName,
      params,
      sourceId: opts?.sourceId,
    });

    if (result.action === Decision.BLOCK) {
      throw new ToolCallBlocked(toolName, result.reason, result.trace_id);
    }

    if (result.action === Decision.REQUIRE_CONFIRMATION) {
      if (!this.confirmCallback) {
        throw new ConfirmationRejected(toolName);
      }
      const confirmed = await this.confirmCallback(toolName, params);
      if (!confirmed) {
        throw new ConfirmationRejected(toolName);
      }
    }

    return fn(params);
  }
}

/**
 * Represents a guarded agent session.
 *
 * Usage:
 * ```ts
 * const session = await shield.session("Summarize my emails");
 * try {
 *   const result = await session.executor.execute("read_inbox", { limit: 10 }, readInbox);
 * } finally {
 *   session.close();
 * }
 * ```
 */
export class ShieldSession {
  public sessionId = "";
  public traceId = "";
  private _executor: GuardedExecutor | null = null;

  constructor(
    private readonly client: ServerClient,
    private readonly userMessage: string,
    private readonly agentId: string,
    private readonly metadata: Record<string, unknown>,
    private readonly confirmCallback?: ConfirmCallback,
  ) {}

  /** Initialize the session by registering it with the server. */
  async open(): Promise<this> {
    const info = await this.client.createSession({
      userMessage: this.userMessage,
      agentId: this.agentId,
      metadata: this.metadata,
    });
    this.sessionId = info.session_id;
    this.traceId = info.trace_id;
    this._executor = new GuardedExecutor(
      this.client,
      this.sessionId,
      this.confirmCallback,
    );
    return this;
  }

  /** Close the session. Currently a no-op as lifecycle is server-tracked. */
  close(): void {
    // Session lifecycle is tracked server-side.
  }

  /** Return the executor that routes tool calls through the server check. */
  get executor(): GuardedExecutor {
    if (!this._executor) {
      throw new Error("ShieldSession.open() must be called before accessing executor");
    }
    return this._executor;
  }
}
